File: backend/internal/domain/stats.go
package domain

import "time"

type StatsSummary struct {
	QueriesTotal   int64
	QueriesBlocked int64
	BlockedPercent float64
	GravitySize    int64
	UniqueClients  int64
	UniqueDomains  int64
	Took           time.Duration
}

type StatsHistoryPoint struct {
	Timestamp time.Time
	Total     int64
	Blocked   int64
}

type StatsHistory struct {
	Points []StatsHistoryPoint
	Took   time.Duration
}

type TopDomainEntry struct {
	Domain string
	Count  int64
}

type StatsTopDomains struct {
	TopQueried []TopDomainEntry
	TopBlocked []TopDomainEntry
	Took       time.Duration
}

type TopClientEntry struct {
	IP    string
	Name  string
	Count int64
}

type StatsTopClients struct {
	Clients []TopClientEntry
	Took    time.Duration
}

type ClusterStatsSummary struct {
	Cluster StatsSummary
	Nodes   map[int64]*NodeResult[*StatsSummary]
}

type ClusterStatsHistory struct {
	ClusterPoints []StatsHistoryPoint
	Nodes         map[int64]*NodeResult[*StatsHistory]
}

type ClusterStatsTopDomains struct {
	ClusterTopQueried []TopDomainEntry
	ClusterTopBlocked []TopDomainEntry
	Nodes             map[int64]*NodeResult[*StatsTopDomains]
}

type ClusterStatsTopClients struct {
	ClusterClients []TopClientEntry
	Nodes          map[int64]*NodeResult[*StatsTopClients]
}

File: frontend/src/types/stats.ts
export type StatsSummary = {
	queriesTotal: number;
	queriesBlocked: number;
	blockedPercent: number;
	gravitySize: number;
	uniqueClients: number;
	uniqueDomains: number;
};

export type StatsHistoryPoint = {
	timestamp: string; // RFC3339
	total: number;
	blocked: number;
};

export type TopDomain = {
	domain: string;
	count: number;
};

export type TopClient = {
	ip: string;
	name: string;
	count: number;
};

export type StatsNode<T> = {
	node: { id: number; name: string; host: string };
	success: boolean;
	error?: string;
	data: T;
};

export type StatsSummaryResponse = {
	cluster: StatsSummary;
	nodes: StatsNode<StatsSummary>[];
};

export type StatsHistoryResponse = {
	cluster: StatsHistoryPoint[];
	nodes: StatsNode<StatsHistoryPoint[]>[];
};

export type StatsTopDomainsNodePayload = {
	topQueried: TopDomain[];
	topBlocked: TopDomain[];
};

export type StatsTopDomainsResponse = {
	clusterTopQueried: TopDomain[];
	clusterTopBlocked: TopDomain[];
	nodes: StatsNode<StatsTopDomainsNodePayload>[];
};

export type StatsTopClientsResponse = {
	clusterClients: TopClient[];
	nodes: StatsNode<TopClient[]>[];
};

export type StatsRange = '1h' | '6h' | '24h';
